Add Cryptozoologist.state method

andrewmbyrd opened the pull request:
Fixes #15 

@feministy, I've got a working implementation of what I think fits the desired output.
I haven't implemented spec tests yet, nor have I updated README etc., I just wanted to get your take first.

I've added all 50 U.S. states to a new `states` Dictionary.  The `Generator.state` method takes a random state from that list. If the selected state has two words, it will replace (randomly) one of the words with an alliterative match from another library (currently `:animal, :clothing, or :colors`). So `New Hampshire` might turn into `Newt Hampshire` or `New Honey bee`

If the selected state is only one word, then we select a random letter from index 0  to `state_name.length - 1` and replace that with an alliterative word from another library.

There is logic to determine how to replace the original state name depending on how long the name from the other library is.

So if I have "Idaho" and the random `key_letter` is the "d", i could end up with "I_dog_aho" because dog isn't longer than the letters "daho". 

If the word from the other library comes back as, say, "dragonfly", then I'll get "Idragonfly" because "dragonfly" _is_ longer than "daho".

My favorite ones so far are "Californiguana" and "Wiscouterwear" :smile: 

I've left some `puts` statements to make it easy to quickly see whats going on in the console.

Please let me know your feedback, and I can tidy up the code and add tests from there!

feministy commented:
i am going to try and review this tonight but i may not be able to get it to it until tomorrow. this is super awesome by the way - i love your take on the two word state replacements 😻

feministy commented:
i did not forget about you, i promise!

feministy commented on README.md:
```diff
@@ -22,12 +22,15 @@ Or install it yourself as:
 * `Cryptozoologist.lorem(sentence_count)`: returns `sentence_count` number of sentences, separated by punctuation randomly selected from `["!", ".", "?"]`
 * `Cryptozoologist.street_address`: returns a string formatted to US street address standards (house number and street name) using the animal dictionaries (_Note_: ignores animal subdictionary exclusions)
 
+* `Cryptozoologist.state`: returns a string of a state name combined with a random word from other internal dictionaries.
```
i think there's an extra blank line between this bullet and the previous one that shouldn't be there.

feministy commented on README.md:
```diff
@@ -22,12 +22,15 @@ Or install it yourself as:
 * `Cryptozoologist.street_address`: returns a string formatted to US street address standards (house number and street name) using the animal dictionaries (_Note_: ignores animal subdictionary exclusions)
 
+* `Cryptozoologist.state`: returns a string of a state name combined with a random word from other internal dictionaries.
+(_Note_: ignores dictionary exclusions) 
```
this can be moved up onto the same line like the previous bullet

feministy commented on lib/cryptozoologist/generator.rb:
```diff
@@ -48,12 +48,49 @@ def random
+        else
+          base_word[key_letter_index] = insert_word
+          final_word = base_word
+        end
```
the nested `if` statements are difficult for me to parse, would it be possible to refactor the top level conditional into separate methods to help make this easier to read?

feministy commented on lib/cryptozoologist/generator.rb:
```diff
@@ -48,12 +48,49 @@ def random
+        base_word = base_state_words[0]
+        key_letter_index = rand(0...base_word.length-1)
+        insert_word = get_alliteration(libraries, base_word[key_letter_index]).downcase
+        if insert_word.length > base_word.length - key_letter_index && key_letter_index > 0
```
for things like this that are the heart of the algorithm, a comment explaining what is happening is super helpful for me!

andrewmbyrd commented:
thanks for the feedback! I've refactored the state method so that it will make a different (private) method call depending on how many  words the state name has, and I've added comments to each method. Let me know your thoughts :smile:

feministy commented on lib/cryptozoologist/generator.rb:
```diff
@@ -48,12 +48,71 @@ def random
+        final_word = handle_two_word(state_name, libraries)
+      else
+        final_word = handle_one_word(state_name, libraries)
+      end
```
we can just end with the conditional here, we don't need the variable 

```ruby
if has_two_words
  handle_two_word(state_name, libraries)
else
  handle_one_word(state_name, libraries)
end
```

andrewmbyrd replied:
ah right! thanks I've got that updated

feministy commented on lib/cryptozoologist/generator.rb:
```diff
@@ -48,12 +48,71 @@ def random
+    def state
+      libraries = [:animals, :clothing, :colors]
+      state_name = Dictionary.states.sample
+      has_two_words = state_name.index(" ")
```
smart!!

feministy commented on lib/cryptozoologist/generator.rb:
```diff
@@ -48,12 +48,71 @@ def random
+
+      #if the randomly selected word from libraries has more letters than the substring from key_letter_index to state_name.length
+      #completely replace that substring
+      if insert_word.length > state_name.length - key_letter_index && key_letter_index > 0
```
can you add some parenthesis around this? it's difficult for me to discern the order of operations here with all the comparisons and mathing

andrewmbyrd replied:
good point, I've put some parentheses in there

andrewmbyrd commented:
@feministy I'm curious to hear how best to implement tests for this? Given that the returned string may be one word or two, and that the random word may be anywhere in the return word, all we know for sure is that we should have a string returned to us: 

```
context '#state' do
    before do
      @state = Cryptozoologist.state
    end

    it 'returns a string' do
      expect(@state).to be_instance_of(String)
    end
end
```

I could make the private methods into public methods in order to test them more directly, but what's your take on it?

feministy commented:
@andrewmbyrd a strategy i might use in this scenario would be to stub out the dictionaries to test the one and two word replacement functionality.

for the state dictionary and _one_ of the dictionaries used by the replacement logic, i would include only one word each. the other dictionaries could be stubbed to be empty - this would make it easier for you to verify that it works as expected.

you'd have to stub for several test cases:

- one word state
- two word state
- state that's longer than the replacement
- state that's shorter than the replacement
- etc ...???

does that make sense? basically you want to control the setting so the test returns what you expect the logic in the code to do, and the easiest way to do that is to stub the dictionaries themselves.

feministy commented:
just a heads up i changed this to the branch for v3 instead of master! no pressure to get changes in before v3 rolls out, i can always add this to a future version 👍

andrewmbyrd commented:
@feministy , thanks for reminding me to get rolling on this!
So with the `state` function, there were 3 degrees of freedom. One for the state name, one for the library, and one for the replacement index of the letter in the state name. In order to produce perfectly repeatable tests, I've added 3 optional arguments to the state function for each of those degrees of freedom.

Now the tests all pass with the same results every time. Let me know what you think!
